repo-scripts/size-analysis/analysis-helper.ts

```diff
@@ -0,0 +1,316 @@
+ * of the minified build.
+ */
+export async function extractDependenciesAndSize(
+  exportName: string,
+  jsBundle: string,
+  map: Map<string, string>
+): Promise<ExportData> {
+  const input = tmp.fileSync().name + '.js';
+  const output = tmp.fileSync().name + '.js';
+
+  const beforeContent = `export { ${exportName} } from '${path.resolve(
+    jsBundle
+  )}';`;
+
+  fs.writeFileSync(input, beforeContent);
+
+  // Run Rollup on the JavaScript above to produce a tree-shaken build
+  const bundle = await rollup.rollup({
+    input,
+    plugins: [
+      resolve({
+        mainFields: ['esm2017', 'module', 'main']
+      }),
+      commonjs()
+    ]
+  });
+  await bundle.write({ file: output, format: 'es' });
+
+  const dependencies = extractDeclarations(output, map);
+
+  // Extract size of minified build
+  const afterContent = fs.readFileSync(output, 'utf-8');
+
+  const { code } = terser.minify(afterContent, {
+    output: {
+      comments: false
+    },
+    mangle: true,
+    compress: false
+  });
+
+  fs.unlinkSync(input);
+  fs.unlinkSync(output);
+  return { dependencies, sizeInBytes: Buffer.byteLength(code!, 'utf-8') };
+}
+
+/**
+ * Extracts all function, class and variable declarations using the TypeScript
+ * compiler API.
+ */
+export function extractDeclarations(
+  jsFile: string,
+  map?: Map<string, string>
+): MemberList {
+  const program = ts.createProgram([jsFile], { allowJs: true });
+  const checker = program.getTypeChecker();
+
+  const sourceFile = program.getSourceFile(jsFile);
+  if (!sourceFile) {
+    throw new Error('Failed to parse file: ' + jsFile);
+  }
+
+  let declarations: MemberList = {
+    functions: [],
+    classes: [],
+    variables: [],
+    enums: []
+  };
+
+  ts.forEachChild(sourceFile, node => {
+    if (ts.isFunctionDeclaration(node)) {
+      declarations.functions.push(node.name!.text);
+    } else if (ts.isClassDeclaration(node)) {
+      declarations.classes.push(node.name!.text);
+    } else if (ts.isVariableDeclaration(node)) {
+      declarations.variables.push(node.name!.getText());
+    } else if (ts.isEnumDeclaration(node)) {
+      declarations.enums.push(node.name.escapedText.toString());
+    } else if (ts.isVariableStatement(node)) {
+      const variableDeclarations = node.declarationList.declarations;
+      variableDeclarations.forEach(variableDeclaration => {
+        if (ts.isIdentifier(variableDeclaration.name)) {
+          declarations.variables.push(
+            (variableDeclaration.name as ts.Identifier).getText(sourceFile)
+          );
+        }
+        // TODO: variableDeclaration.name is an union type (Identifier | BindingPattern)
+        // need Identifier type, not sure in what case BindingPattern type is for.
+        else {
+          console.log(
+            'this VariableDeclaration.name object is of BindingPattern type !'
+          );
+        }
+      });
+    } else if (ts.isExportDeclaration(node)) {
```

Comment:
This block has some complex logic that will be hard to understand for people who is not familiar with ts compiler and the nuances in exports. We should add more comments explaining the code assuming the reader doesn't have these knowledge, and make it clear to them.